### PR TITLE
try to turn on all WildFly statistics if monitoring a WildFly server

### PR DESCRIPTION
**Pull request by jmazzitelli**
this should be ready

### CODE REVIEW

**Comment by jmazzitelli** [comment]
This will only work in standalone mode. You cannot change config for slave servers. See Brian's comment here:

https://issues.jboss.org/browse/WFCORE-771?focusedCommentId=13199892#comment-13199892

And I see that error message when my code tries to change the statistics enable flags on slave servers. So that matches what Brian says should happen.

**Comment by jmazzitelli** [comment]
I believe this is ready.

In short, the agent upon start up will now attempt to turn on the different subsystems' statistics enablement flags. This is the default behavior, but you can turn it off by setting "enable-statistics: false" in the local-dmr or remote-dmr section under the managed-server section.

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/protocol/dmr/StatisticsControl.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            } else {
+                for (String host : hosts) {
+                    List<String> servers = getDomainHostServers(mcc, host);
+                    for (String server : servers) {
+                        // I was trying to be too clever. This would work awesomely if WildFly would just allow it.
+                        // However, this error occurs when you try to set attributes on slave servers:
+                        // "User operations are not permitted to directly update the persistent configuration
+                        // of a server in a managed domain."
+                        // So, for now, comment this out until perhaps in the future WildFly/EAP will allow this.
+                        // This means when in domain mode, the user must manually turn on statistics themselves.
```
Comment (jshaughn) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/protocol/dmr/StatisticsControl.java:63]: Wouldn't the slaves be enabled if you enable the controller?

**Review by jshaughn — APPROVED** [review]
This seems to be in good shape.  I guess my question is, "Why we are doing this?".   And if we enable stats by default does this have any perf effect?

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/protocol/dmr/StatisticsControl.java (hunk 1 above)
Comment (jmazzitelli) [reply hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/protocol/dmr/StatisticsControl.java:63]: no, the slaves configuration here are not affected by changes in the host controller (in fact, for example, there is no /subsystem=ejb3 in the host controller anyway.

**Comment by jmazzitelli** [comment]
> I guess my question is, "Why we are doing this?"

We are doing this because:

1. By default WildFly does not enable statistics and in this state every metric that Prometheus exporter will collect and scrape for these subsystems will always be 0

2. One of our main contributions to middleware management is to provide monitoring (metrics and graphing) - we are rendered mostly useless if a lot of our subsystem metrics are always 0.

3. Because of this, today we tell users they have to enable these stat flags manually. We do provide a cmdgw command for them to execute (StatisticsCommand), but it requires MiQ to send that message and for the user to tell MiQ to send that message. Someone may start Hawkular and not see many metrics for these subsystems and will not know why. This PR makes things more automatic.

> And if we enable stats by default does this have any perf effect?

I will assume it will affect performance but it is unknown if it is a lot or  a little. I suspect it isn't much. But in any case, this is why I provide "enable-statistics" flag in the agent configuration - you can turn this off if you find the performance hit is not worth it. In fact, by default I set this to true, we could set the default to "false" and just have the user pass in a env var to turn it on if they want.

let me know if we want to change the default to false and have it be enabled by a system property and I can do that. That is simple to do. I just figured 'we are a system that collects metrics - why would we not want to enable metrics by default?

**Comment by jshaughn** [comment]
This was sufficient: "By default WildFly does not enable statistics and in this state every metric that Prometheus exporter will collect and scrape for these subsystems will always be 0".

**Pull request merged**